### PR TITLE
upgrade to nan v2.2.0.

### PR DESCRIPTION
**Pull request by chjj**
Needs a review. It builds and runs. ~~Unfortunately, lmdb is throwing an error after 10-20k writes: `MDB_MAP_FULL: Environment mapsize limit reached`~~ (wasn't setting the map size).

cc @rvagg @ralphtheninja

### CODE REVIEW

**Comment by chjj** [comment]
And it looks like I somehow created a memory leak. :) JS heap is getting pretty big. Investigating.

File: src/database.cc [hunk 1]
```diff
@@ -323,85 +323,85 @@ v8::Handle<v8::Value> Database::NewInstance (v8::Local<v8::String> &location) {
 }
 
 NAN_METHOD(Database::Open) {
-  NanScope();
+  Nan::HandleScope scope;
```
Comment (ralphtheninja) [inline src/database.cc]: iirc you can throw these old `NanScope()` away

File: src/batch.cc [hunk 2]
```diff
@@ -114,35 +114,35 @@ void WriteBatch::Clear () {
   }
 }
 
-static v8::Persistent<v8::FunctionTemplate> writebatch_constructor;
+static Nan::Persistent<v8::FunctionTemplate> writebatch_constructor;
 
 void WriteBatch::Init () {
-  NanScope();
+  Nan::HandleScope scope;
```
Comment (ralphtheninja) [inline src/batch.cc]: This should be ok since we're not in a `NAN_METHOD()`

File: src/database.cc (hunk 1 above)
Comment (chjj) [inline src/database.cc]: Will do.

**Comment by chjj** [comment]
I cannot figure out what's leaking on the js heap. Spent a while debugging it yesterday. I figured it would be a missing HandleScope somewhere, but I ruled out pretty much everything. Maybe something to do with how persistent handles are registered? Not sure.

Like some cruel joke, I get away from one memory problem only to bump into another.

**Comment by chjj** [comment]
I seem to have fixed the memory leak somewhere in all my refactoring. Works well now.

File: src/database.cc [hunk 3]
```diff
@@ -426,36 +418,34 @@ NAN_METHOD(Database::Close) {
...
           }
         }
     }
   } else {
-    NanAsyncQueueWorker(worker);
+    Nan::AsyncQueueWorker(worker);
   }
 
-  NanReturnUndefined();
+  info.GetReturnValue().SetUndefined();
```
Comment (rvagg) [inline src/database.cc]: can probably remove this completely

File: src/database.cc [hunk 4]
```diff
@@ -463,78 +453,73 @@ NAN_METHOD(Database::Put) {
...
     , key
     , value
     , keyHandle
     , valueHandle
   );
-  NanAsyncQueueWorker(worker);
+  Nan::AsyncQueueWorker(worker);
 
-  NanReturnUndefined();
+  info.GetReturnValue().SetUndefined();
```
Comment (rvagg) [inline src/database.cc]: ditto

File: src/database.cc [hunk 5]
```diff
@@ -463,78 +453,73 @@ NAN_METHOD(Database::Put) {
...
+    , new Nan::Callback(callback)
     , key
     , asBuffer
     , keyHandle
   );
-  NanAsyncQueueWorker(worker);
+  Nan::AsyncQueueWorker(worker);
 
-  NanReturnUndefined();
+  info.GetReturnValue().SetUndefined();
```
Comment (rvagg) [inline src/database.cc]: ditto

File: src/database.cc [hunk 6]
```diff
@@ -463,78 +453,73 @@ NAN_METHOD(Database::Put) {
...
-    , new NanCallback(callback)
+    , new Nan::Callback(callback)
     , key
     , keyHandle
   );
-  NanAsyncQueueWorker(worker);
+  Nan::AsyncQueueWorker(worker);
 
-  NanReturnUndefined();
+  info.GetReturnValue().SetUndefined();
```
Comment (rvagg) [inline src/database.cc]: ...

File: src/database.cc [hunk 7]
```diff
@@ -563,39 +548,37 @@ NAN_METHOD(Database::Batch) {
...
   v8::Handle<v8::Object> iteratorHandle = Iterator::NewInstance(
-      args.This()
-    , v8::Number::New(id)
+      info.This()
+    , Nan::New<v8::Number>(id)
     , optionsObj
   );
   if (try_catch.HasCaught()) {
-    node::FatalException(try_catch);
+    Nan::FatalException(try_catch);
```
Comment (rvagg) [inline src/database.cc]: this may need an early return, not sure

File: src/nlmdb.cc [hunk 8]
```diff
@@ -9,17 +9,17 @@
 
 namespace nlmdb {
 
-void Init (v8::Handle<v8::Object> target) {
+NAN_MODULE_INIT(init) {
```
Comment (rvagg) [inline src/nlmdb.cc]: this could probably do with a `Nan:HandleScope scope;`

File: src/nlmdb.cc [hunk 9]
```diff
@@ -9,17 +9,17 @@
...
 
   v8::Local<v8::Function> nlmdb =
-      v8::FunctionTemplate::New(NLMDB)->GetFunction();
+      Nan::New<v8::FunctionTemplate>(NLMDB)->GetFunction();
 
-  target->Set(v8::String::NewSymbol("nlmdb"), nlmdb);
+  target->Set(Nan::New("nlmdb").ToLocalChecked(), nlmdb);
 }
 
-NODE_MODULE(nlmdb, Init)
```
Comment (rvagg) [inline src/nlmdb.cc]: fwiw name isn't important here, could just as well be `Init` and you could also avoid `NAN_MODULE_INIT` by changing the `Handle` to a `Local` in the original `Init` declaration. No big deal either way though.

File: src/nlmdb.h [hunk 10]
```diff
@@ -41,6 +41,17 @@ static inline bool BooleanOptionValue(
     && optionsObj->Get(opt)->BooleanValue();
 }
 
+static inline bool BooleanOptionValueDef(
```
Comment (rvagg) [inline src/nlmdb.h]: I honestly don't know if it's necessary but I think we should put a `Nan::HandleScope scope;` in here and the three following functions. They are `inline` but I don't know if that provides any guarantees and adding an extra `HandleScope` isn't going to hurt even if it is unnecessary.

**Comment by rvagg** [comment]
This all looks fine to me, just some notes about unnecessary `undefined` returns we can remove now and also defensive `HandleScope`s in the inline utility functions just to be sure that's not a source of leaks.

**Comment by 0x00A** [comment]
Any reason not to merge this?

**Comment by 0x00A** [comment]
Ah, I see why...

```
317 passing (40s)
1 pending
41 failing
```

**Comment by chjj** [comment]
Closing in favor of #14

**Pull request closed without merging**